src/Magento/FunctionalTestingFramework/StaticCheck/TestDependencyCheck.php

```diff
@@ -0,0 +1,296 @@
+{
+    /**
+     * Array of FullModuleName => [dependencies]
+     * @var array
+     */
+    private $allDependencies;
+
+    /**
+     * Array of FullModuleName => [dependencies], including flattened dependency tree
+     * @var array
+     */
+    private $flattenedDependencies;
+
+    /**
+     * Array of FullModuleName => PathToModule
+     * @var array
+     */
+    private $moduleNameToPath;
+
+    /**
+     * Array of FullModuleName => ComposerModuleName
+     * @var array
+     */
+    private $moduleNameToComposerName;
+
+    /**
+     * Transactional Array to keep track of what dependencies have already been extracted.
+     * @var array
+     */
+    private $alreadyExtractedDependencies;
+
+    /**
+     * Checks test dependencies, determined by references in tests versus the dependencies listed in the Magento module
+     *
+     * @param InputInterface $input
+     * @return string
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function execute(InputInterface $input)
+    {
+        MftfApplicationConfig::create(
+            true,
+            MftfApplicationConfig::UNIT_TEST_PHASE,
+            false,
+            false
+        );
+
+        $testObjects = TestObjectHandler::getInstance()->getAllObjects();
+        if (!class_exists('\Magento\Framework\Component\ComponentRegistrar')) {
+            return "TEST DEPENDENCY CHECK ABORTED: MFTF must be attached or pointing to Magento codebase.";
+        }
+        $registrar = new \Magento\Framework\Component\ComponentRegistrar();
+        $this->moduleNameToPath = $registrar->getPaths(\Magento\Framework\Component\ComponentRegistrar::MODULE);
+        $this->moduleNameToComposerName = $this->buildModuleNameToComposerName($this->moduleNameToPath);
+        $this->flattenedDependencies = $this->buildComposerDependencyList();
+
+        $testErrors = [];
+        foreach ($testObjects as $testObject) {
+            // Find testobject's module
+            $allFiles = explode(",", $testObject->getFilename());
+            $basefile = $allFiles[0];
+            $modulePath = dirname(dirname(dirname(dirname($basefile))));
+            $moduleFullName = array_search($modulePath, $this->moduleNameToPath) ?? null;
+
+            // Not a module, is either dev/tests/acceptance or loose folder with test materials
+            if ($moduleFullName == null) {
+                continue;
+            }
+
+            // Force Test to resolve references
+            $testObject->getOrderedActions();
+
+            // Find objects test referenced
+            $dataReferences = DataObjectHandler::getInstance()->getAccessedObjects();
+            $sectionReferences = SectionObjectHandler::getInstance()->getAccessedObjects();
+            $pageReferences = PageObjectHandler::getInstance()->getAccessedObjects();
+            $actionGroupReferences = ActionGroupObjectHandler::getInstance()->getAccessedObjects();
+
+            // If test extends, find reference
```

Comment:
By the definition of the static check's goal, that test should fail since it violates the composer dependencies of the Magento module.
I do understand the usecase (sort of), we could implement a test attribute `ignoreDependencies` though for this sort of thing (kind like `@SupressWarnings` for PHPMD).
Again though, that pretty much goes against the whole point of this static test, so I think we should get team feedback/direction on what to do here.

Reply:
No, that's not what I mean.
Those tests are written specifically to follow dependency rules. I suspect that the static check somehow could not resolve this use case correctly. Because we will suggest others to use similar strategy to write tests to avoid dependency problem. We cannot contradict ourselves.